### PR TITLE
Added filterbanking

### PR DESCRIPTION
**Pull request by gemeinl**
Filterbanking can now be used similarly to the other preprocessing steps.
It modifies a ``mne.io.Raw`` in-place and adds new channels with the filtered signals.

### CODE REVIEW

File: braindecode/datautil/preprocess.py [hunk 1]
```diff
@@ -251,3 +252,54 @@ def scale(data, factor):
...
+               fir_design='firwin', **filter_kwargs):
+    """Applies multiple bandpass filters to the signals in raw. Raw will be
+    modified in-place and number of channels in raw will be updated to
+    len(frequency_bands) * len(raw.ch_names) (-len(raw.ch_names) if
+    drop_original_signals). For filtering arguments please refer to
+    mne.filter.filter_data().
+
+    Params
+    ------
+    raw: mne.Raw
```
Comment (agramfort) [inline braindecode/datautil/preprocess.py:268]:
```suggestion
    raw: instance of mne.io.Raw
```

File: braindecode/datautil/preprocess.py [hunk 2]
```diff
@@ -251,3 +252,54 @@ def scale(data, factor):
...
+        The raw singnals to be filtered
```
Comment (agramfort) [inline braindecode/datautil/preprocess.py:269]:
```suggestion
        The raw signals to be filtered
```

File: braindecode/datautil/preprocess.py [hunk 3]
```diff
@@ -251,3 +252,54 @@ def scale(data, factor):
...
+def filterbank(raw, frequency_bands, drop_original_signals=True, n_jobs=1,
+               method='fir', iir_params=None, phase='zero', fir_window='hamming',
+               fir_design='firwin', **filter_kwargs):
+    """Applies multiple bandpass filters to the signals in raw. Raw will be
+    modified in-place and number of channels in raw will be updated to
+    len(frequency_bands) * len(raw.ch_names) (-len(raw.ch_names) if
+    drop_original_signals). For filtering arguments please refer to
+    mne.filter.filter_data().
+
+    Params
```
Comment (agramfort) [inline braindecode/datautil/preprocess.py:266]: Parameters

is the standard for numpydoc

File: braindecode/datautil/preprocess.py [hunk 4]
```diff
@@ -251,3 +252,54 @@ def scale(data, factor):
...
+    ------
+    raw: mne.Raw
+        The raw singnals to be filtered
+    frequency_bands: list(tuple)
+        The frequency bands to be fitlered for (e.g. [(4, 8), (8, 13)])
+    drop_original_signals: bool
+        Whether to drop the original unfiltered signals
```
Comment (agramfort) [inline braindecode/datautil/preprocess.py:273]: missing many params

Comment (gemeinl) [reply braindecode/datautil/preprocess.py:273]: In the docstring above I stated to refer to ```mne.filter.filter_data()``` for a description of the parameters. Should I copy the docstring from mne instead and paste it here?

**Comment by agramfort** [comment]
I would say yes. I expect to learn about parameters in the Parameters
section. Even **filter_kwargs should be explained there.

>

File: braindecode/datautil/preprocess.py [hunk 5]
```diff
@@ -251,3 +253,63 @@ def scale(data, factor):
...
+        for ch in raw.ch_names[i::len(original_ch_names)]]
+    raw.reorder_channels(chs_by_freq_band)
+    if drop_original_signals:
+        raw.drop_channels(original_ch_names)
+
+
+# copy docstring from mne to here
+filterbank.__doc__ += mne.io.Raw.filter.__doc__[
+        mne.io.Raw.filter.__doc__.find("picks : "):
+        mne.io.Raw.filter.__doc__.find("Returns")].replace(8 * " ", 4 * " ")
```
Comment (agramfort) [inline braindecode/datautil/preprocess.py:315]: Honestly this is black magic. This code in mne barely changes so I would copy the docstring.
YOu can make a test that asserts that docs are in sync. It's much simpler.

Comment (robintibor) [reply braindecode/datautil/preprocess.py:315]: I feel previous solution to explicitly refer to MNE also has advantages:
* Less clutter in our code
* It is more explicit that we just forward these arguments, one does not have to check whether there is any difference between MNE and us

And we anyways build so much around MNE now, pretty much encouraging the user to learn some MNE.

File: braindecode/datautil/preprocess.py [hunk 6]
```diff
@@ -251,3 +253,63 @@ def scale(data, factor):
     if hasattr(data, '_data'):
         data._data = scaled
     return scaled
+
+
+def filterbank(raw, frequency_bands, drop_original_signals=True, n_jobs=1,
+               method='fir', iir_params=None, phase='zero', fir_window='hamming',
+               fir_design='firwin', **filter_kwargs):
```
Comment (agramfort) [inline braindecode/datautil/preprocess.py:260]: why not simply having **filter_kwargs and say it supports params of filter function from mne?

if feels you explicit some params and some not.

Comment (gemeinl) [reply braindecode/datautil/preprocess.py:260]: In another PR you mentioned disliking kwargs and to make arguments explicit. I therefore took the arguments that are relevant for my use case and made them explicit, while at the same time keeping full compatibility with mne through filter_kwargs. I will now update it again to be less explicit and only have the filter_kwargs. The docstring for this parameter will then contain a referral to mne.

Comment (agramfort) [reply braindecode/datautil/preprocess.py]: arfff to be it's one option or the other. If it was me and I would copy all
the mne parameters and copy the docstring. After updating the docstring
5 times due to mne changes I would reconsider. Note that mne release
are every 6 months so...

**Pull request merged**